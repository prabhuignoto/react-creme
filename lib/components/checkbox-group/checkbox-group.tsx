import classNames from "classnames";
import { nanoid } from "nanoid";
import React, { useCallback, useEffect, useMemo } from "react";
import { CheckBox } from "../checkbox/checkbox";
import { CheckboxModel } from "../checkbox/checkbox-model";
import { useFirstRender } from "../common/effects/useFirstRender";
import "./checkbox-group.scss";

interface CheckboxGroupProps {
  options: CheckboxModel[];
  border?: boolean;
  disabled?: boolean;
  layout?: "horizontal" | "vertical";
  checkboxStyle?: "square" | "round";
  onChange?: (
    selected: { id?: string; name?: string; isChecked: boolean }[]
  ) => void;
  noUniqueIds?: boolean;
}

const CheckBoxGroup: React.FunctionComponent<CheckboxGroupProps> = ({
  options = [],
  disabled,
  border,
  layout = "vertical",
  checkboxStyle = "square",
  onChange,
  noUniqueIds = false,
}) => {
  const wrapperClass = useMemo(() => {
    return classNames("rc-checkbox-group", {
      "rc-checkbox-group-border": border,
      "rc-checkbox-group-vertical": layout === "vertical",
      "rc-checkbox-group-horizontal": layout === "horizontal",
      "rc-checkbox-group-disabled": disabled,
    });
  }, [layout, disabled]);

  const [items, setItems] = React.useState(
    options
      ? options.map((option) => ({
          ...(!noUniqueIds ? { id: nanoid() } : { id: option.id }),
          ...option,
          isChecked: option.isChecked || false,
        }))
      : []
  );

  const handleChange = useCallback(
    (id: string, name: string, selected: boolean) => {
      setItems((items) => {
        const _newItems = items.map((item) => {
          return {
            ...item,
            isChecked: item.id === id ? selected : !!item.isChecked,
          };
        });

        return _newItems;
      });
    },
    []
  );

  useEffect(() => {
    if (!isFirstRender.current) {
      onChange?.(
        items.map((item) => ({
          id: item.id,
          name: item.label,
          isChecked: item.isChecked,
        }))
      );
    }
  }, [JSON.stringify(items.map((i) => i.isChecked))]);

  const isFirstRender = useFirstRender();

  return (
    <div className={wrapperClass} role="group">
      {items.map((item) => {
        const { id, ...rest } = item;
        return (
          <CheckBox
            key={id}
            {...rest}
            disabled={disabled || item.disabled}
            onChange={handleChange}
            checkBoxStyle={checkboxStyle}
            noUniqueId={true}
            id={item.id}
          />
        );
      })}
    </div>
  );
};

export { CheckBoxGroup };
