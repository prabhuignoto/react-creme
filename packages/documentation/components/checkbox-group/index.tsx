import React, { lazy } from "react";
import DemoPageRenderer from "../../common/demo-page-renderer";

function index() {
  return (
    <DemoPageRenderer
      title="Checkbox Group"
      description="Checkbox group is a group of checkboxes that can be used to select multiple options."
      demoWidget={lazy(() => import("./widgets"))}
      callbacks={[
        {
          name: "onChange",
          description:
            "Callback function to be called when the checkbox group changes.",
          default: "",
          optional: "Yes",
          type: "function",
        },
      ]}
      properties={[
        {
          name: "options",
          description: "Options to be displayed in the checkbox group.",
          default: "[]",
          optional: "Yes",
          type: "Array<Option>",
        },
        {
          name: "noUniqueIds",
          description:
            "If true, the checkbox group will not generate unique ids for each checkbox.",
          default: "false",
          optional: "Yes",
          type: "boolean",
        },
        {
          name: "checkboxStyle",
          description:
            "checkbox render style. <em>square</em> or <em>round</em>",
          default: "square",
          optional: "Yes",
          type: "string",
        },
        {
          name: "disabled",
          description: "Disables the checkbox group",
          default: "false",
          optional: "Yes",
          type: "boolean",
        },
        {
          name: "layout",
          description: "layout of the checkbox group. vertical or horizontal",
          default: "vertical",
          optional: "Yes",
          type: "string",
        },
      ]}
      tabTitles={["examples", "properties"]}
    ></DemoPageRenderer>
  );
}

export default index;
