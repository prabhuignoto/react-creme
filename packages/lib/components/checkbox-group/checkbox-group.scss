@use '../../design/core.scss';
@use '../../design/theme.scss';

.rc-checkbox-group-wrapper {
  &.rc-checkbox-group-horizontal {
    @extend %left;
  }

  &.rc-checkbox-group-vertical {
    @extend %col;
  }
}
