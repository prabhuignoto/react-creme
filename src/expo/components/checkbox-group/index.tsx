import React from "react";
import DemoPageRenderer from "../../common/demo-page-renderer";
import Widgets from "./widgets";

function index() {
  return (
    <DemoPageRenderer
      title="Checkbox Group"
      description="Checkbox group is a group of checkboxes that can be used to select multiple options."
      demoWidget={<Widgets />}
      data={[
        {
          name: "<em>options</em>",
          description: "Options to be displayed in the checkbox group.",
          default: "[]",
          optional: "Yes",
          type: "Array<Option>",
        },
        {
          name: "<em>noUniqueIds</em>",
          description:
            "If true, the checkbox group will not generate unique ids for each checkbox.",
          default: "<em>false</em>",
          optional: "Yes",
          type: "boolean",
        },
        {
          name: "<em>checkboxStyle</em>",
          description:
            "checkbox render style. <em>square</em> or <em>round</em>",
          default: "<em>square</em>",
          optional: "Yes",
          type: "string",
        },
        {
          name: "<em>onChange</em>",
          description:
            "Callback function to be called when the checkbox group changes.",
          default: "",
          optional: "Yes",
          type: "function",
        },
        {
          name: "<em>disabled</em>",
          description: "Disables the checkbox group",
          default: "<em>false</em>",
          optional: "Yes",
          type: "boolean",
        },
        {
          name: "<em>layout</em>",
          description:
            "layout of the checkbox group. <em>vertical</em> or <em>horizontal</em>",
          default: "<em>vertical</em>",
          optional: "Yes",
          type: "string",
        },
      ]}
      tabTitles={["examples", "properties"]}
    ></DemoPageRenderer>
  );
}

export default index;
