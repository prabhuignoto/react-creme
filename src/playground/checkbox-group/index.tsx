import React from "react";
import DemoPageRenderer from "./../common/demo-page-renderer";
import Widgets from "./widgets";

function index() {
  return (
    <DemoPageRenderer
      title="Checkbox Group"
      description="Checkbox group is a group of checkboxes that can be used to select multiple options."
      demoWidget={<Widgets />}
      data={[]}
      tabTitles={["examples", "properties"]}
    ></DemoPageRenderer>
  );
}

export default index;
