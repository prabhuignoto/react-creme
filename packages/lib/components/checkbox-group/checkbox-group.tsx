import classNames from "classnames";
import { nanoid } from "nanoid";
import React, { useCallback, useEffect, useMemo } from "react";
import { CheckBox } from "../checkbox/checkbox";
import { CheckboxModel } from "../checkbox/checkbox-model";
import { useFirstRender } from "../common/effects/useFirstRender";
import "./checkbox-group.scss";

export interface CheckboxGroupProps {
  options: CheckboxModel[];
  border?: boolean;
  disabled?: boolean;
  layout?: "horizontal" | "vertical";
  checkboxStyle?: "square" | "round";
  onChange?: (
    selected: { id?: string; name?: string; isChecked: boolean }[]
  ) => void;
  noUniqueIds?: boolean;
  RTL?: boolean;
}

const CheckBoxGroup: React.FunctionComponent<CheckboxGroupProps> = ({
  options = [],
  disabled,
  border,
  layout = "vertical",
  checkboxStyle = "square",
  onChange,
  noUniqueIds = false,
  RTL = false,
}) => {
  const wrapperClass = useMemo(() => {
    return classNames("rc-checkbox-group", {
      "rc-checkbox-group-border": border,
      "rc-checkbox-group-disabled": disabled,
      "rc-checkbox-group-horizontal": layout === "horizontal",
      "rc-checkbox-group-vertical": layout === "vertical",
    });
  }, [layout, disabled]);

  const [items, setItems] = React.useState(
    options
      ? options.map((option) => ({
          ...(!noUniqueIds ? { id: nanoid() } : { id: option.id }),
          ...option,
          isChecked: option.isChecked || false,
        }))
      : []
  );

  const handleChange = useCallback(
    (id: string, name: string, selected: boolean) => {
      setItems((items) => {
        const _newItems = items.map((item) => {
          return {
            ...item,
            isChecked: item.id === id ? selected : !!item.isChecked,
          };
        });

        return _newItems;
      });
    },
    []
  );

  useEffect(() => {
    if (!isFirstRender.current) {
      onChange?.(
        items.map((item) => ({
          id: item.id,
          isChecked: item.isChecked,
          name: item.label,
        }))
      );
    }
  }, [JSON.stringify(items.map((i) => i.isChecked))]);

  const isFirstRender = useFirstRender();

  return (
    <div className={wrapperClass} role="group">
      {items.map((item) => {
        const { id, ...rest } = item;
        return (
          <CheckBox
            key={id}
            {...rest}
            disabled={disabled || item.disabled}
            onChange={handleChange}
            checkBoxStyle={checkboxStyle}
            noUniqueId={true}
            id={item.id}
            RTL={RTL}
          />
        );
      })}
    </div>
  );
};

export { CheckBoxGroup };
