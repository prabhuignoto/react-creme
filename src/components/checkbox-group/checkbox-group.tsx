import classNames from "classnames";
import { nanoid } from "nanoid";
import React, { useMemo } from "react";
import { CheckBox } from "../checkbox/checkbox";
import { CheckboxModel } from "../checkbox/checkbox-model";
import "./checkbox-group.scss";

interface CheckboxGroupProps {
  options: CheckboxModel[];
  border?: boolean;
  disabled?: boolean;
  layout?: "horizontal" | "vertical";
}

const CheckBoxGroup: React.FunctionComponent<CheckboxGroupProps> = ({
  options,
  disabled,
  border,
  layout = "horizontal",
}) => {
  const wrapperClass = useMemo(() => {
    return classNames("rc-checkbox-group", {
      "rc-checkbox-group-border": border,
      "rc-checkbox-group-vertical": layout === "vertical",
      "rc-checkbox-group-horizontal": layout === "horizontal",
      "rc-checkbox-group-disabled": disabled,
    });
  }, [layout, disabled]);

  const [items, setItems] = React.useState(
    options ? options.map((option) => ({ id: nanoid(), ...option })) : []
  );

  const handleChange = (id: string, name: string, selected: boolean) => {
    const newItems = items.map((item) => {
      if (item.id === id) {
        return { ...item, isChecked: selected };
      }
      return item;
    });
    setItems(newItems);
  };

  return (
    <div className={wrapperClass} role="group">
      {items.map((item) => {
        const { id, ...rest } = item;
        return (
          <CheckBox
            key={id}
            {...rest}
            disabled={disabled || item.disabled}
            onChange={handleChange}
          />
        );
      })}
    </div>
  );
};

export { CheckBoxGroup };
