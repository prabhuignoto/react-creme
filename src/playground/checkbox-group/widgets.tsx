import React from "react";
import { CheckBoxGroup, Section } from "../../components";
import { DemoWidget } from "../common/demo-widget";

function Widgets() {
  return (
    <div className="rc-demo-widgets">
      <Section title="Default Render">
        <DemoWidget>
          <CheckBoxGroup
            onChange={(value) => console.log(value)}
            options={[
              {
                label: "Option 1",
              },
              {
                label: "Option 2",
              },
              {
                label: "Option 3",
              },
              {
                label: "Option 4",
              },
            ]}
          />
        </DemoWidget>
      </Section>
      <Section title="Preselected Option">
        <DemoWidget>
          <CheckBoxGroup
            options={[
              {
                label: "Option 1",
              },
              {
                label: "Option 2",
              },
              {
                label: "This option is preselected",
                isChecked: true,
              },
            ]}
          />
        </DemoWidget>
      </Section>
      <Section title="Disabled Option">
        <DemoWidget>
          <CheckBoxGroup
            options={[
              {
                label: "Option 1",
              },
              {
                label: "This option is disabled",
                disabled: true,
              },
              {
                label: "This option is disabled but preselected",
                isChecked: true,
                disabled: true,
              },
            ]}
          />
        </DemoWidget>
      </Section>
      <Section title="Checkbox Group - Round style">
        <DemoWidget>
          <CheckBoxGroup
            checkboxStyle="round"
            options={[
              {
                label: "Option 1",
              },
              {
                label: "Option 2",
              },
              {
                label: "Option 3",
              },
              {
                label: "Option 4",
              },
            ]}
          />
        </DemoWidget>
      </Section>
    </div>
  );
}

export default Widgets;
