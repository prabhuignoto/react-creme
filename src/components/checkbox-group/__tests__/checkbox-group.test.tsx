import { render } from "@testing-library/react";
import React from "react";
import { CheckBoxGroup } from "../checkbox-group";

const options = [
  {
    label: "Option 1",
  },
  {
    label: "Option 2",
    isChecked: true,
  },
  {
    label: "Option 3",
    disabled: true,
  },
];

describe("CheckboxGroup", () => {
  it("should render checkbox group", () => {
    const { getByRole, getByText, getAllByRole } = render(
      <CheckBoxGroup options={options} />
    );

    expect(getByText("Option 1")).toBeInTheDocument();
    expect(getByText("Option 2")).toBeInTheDocument();
    expect(getByText("Option 3")).toBeInTheDocument();

    expect(getByRole("group")).toBeInTheDocument();

    expect(getAllByRole("checkbox")).toHaveLength(3);
  });

  it("should match snapshot", () => {
    const { container } = render(<CheckBoxGroup options={options} />);

    expect(container).toMatchSnapshot();
  });

  it("should check for isChecked", () => {
    const { getAllByRole } = render(<CheckBoxGroup options={options} />);

    expect(getAllByRole("checkbox")[1]).toBeChecked();
  });

  it("check for disabled option", () => {
    const { getAllByRole } = render(<CheckBoxGroup options={options} />);

    expect(getAllByRole("checkbox")[2]).toHaveAttribute(
      "aria-disabled",
      "true"
    );
  });
});
