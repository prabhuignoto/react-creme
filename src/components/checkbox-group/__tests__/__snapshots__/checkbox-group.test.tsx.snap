// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup should match snapshot 1`] = `
<div>
  <div
    class="rc-checkbox-group rc-checkbox-group-horizontal"
    role="group"
  >
    <div
      aria-checked="false"
      aria-labelledby=""
      class="rc-checkbox-wrapper rc-checkbox-sm rc-checkbox-hover rc-checkbox-focus"
      role="checkbox"
      tabindex="0"
    >
      <div
        class="rc-checkbox rc-checkbox-sm"
      >
        <span
          class="rc-checkbox-icon rc-checkbox-sm"
          role="img"
        >
          <svg
            class="prefix__feather prefix__feather-check"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6L9 17l-5-5"
            />
          </svg>
        </span>
      </div>
      <label
        class="rc-checkbox-label rc-checkbox-label-sm"
        id=""
      >
        Option 1
      </label>
    </div>
    <div
      aria-checked="true"
      aria-labelledby=""
      class="rc-checkbox-wrapper rc-checkbox-sm rc-checkbox-hover rc-checkbox-focus"
      role="checkbox"
      tabindex="0"
    >
      <div
        class="rc-checkbox rc-checkbox-sm"
      >
        <span
          class="rc-checkbox-icon rc-checkbox-checked rc-checkbox-sm"
          role="img"
        >
          <svg
            class="prefix__feather prefix__feather-check"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6L9 17l-5-5"
            />
          </svg>
        </span>
      </div>
      <label
        class="rc-checkbox-label rc-checkbox-label-sm"
        id=""
      >
        Option 2
      </label>
    </div>
    <div
      aria-checked="false"
      aria-disabled="true"
      aria-labelledby=""
      class="rc-checkbox-wrapper rc-checkbox-sm rc-disabled rc-checkbox-hover rc-checkbox-focus"
      role="checkbox"
    >
      <div
        class="rc-checkbox rc-disabled rc-checkbox-sm"
      >
        <span
          class="rc-checkbox-icon rc-checkbox-sm"
          role="img"
        >
          <svg
            class="prefix__feather prefix__feather-check"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6L9 17l-5-5"
            />
          </svg>
        </span>
      </div>
      <label
        class="rc-checkbox-label rc-checkbox-label-sm"
        id=""
      >
        Option 3
      </label>
    </div>
  </div>
</div>
`;
